pkg/config/dynamicloader.go

```diff
@@ -0,0 +1,495 @@
+package config
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/fs"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	v2 "github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/agent/pkg/metrics"
+	"github.com/grafana/agent/pkg/metrics/instance"
+	"github.com/grafana/agent/pkg/traces"
+	"github.com/hairyhenderson/go-fsimpl"
+	"github.com/hairyhenderson/go-fsimpl/blobfs"
+	"github.com/hairyhenderson/go-fsimpl/filefs"
+	"github.com/hairyhenderson/gomplate/v3/data"
+	"github.com/hairyhenderson/gomplate/v3/loader"
+	"github.com/hashicorp/go-multierror"
+	"github.com/weaveworks/common/server"
+	"gopkg.in/yaml.v2"
+)
+
+// DynamicLoader is used to load configs from a variety of sources and squash them together.
+// This is used by the dynamic configuration feature to load configurations from a set of templates and then run them through
+// gomplate producing an end result.
+type DynamicLoader struct {
+	loader *loader.ConfigLoader
+	mux    fsimpl.FSMux
+	cfg    *LoaderConfig
+}
+
+// NewDynamicLoader instantiates a new DynamicLoader
+func NewDynamicLoader() (*DynamicLoader, error) {
+
+	return &DynamicLoader{
+		loader: nil,
+		mux:    newFSProvider(),
+		cfg:    nil,
+	}, nil
+}
+
+// LoadConfig loads an already created LoaderConfig into the DynamicLoader
+func (c *DynamicLoader) LoadConfig(cfg LoaderConfig) error {
+	sources := make(map[string]*data.Source)
+	for _, v := range cfg.Sources {
+		sourceURL, err := url.Parse(v.URL)
+		if err != nil {
+			return err
+		}
+		sources[v.Name] = &data.Source{
+			URL:   sourceURL,
+			Alias: v.Name,
+		}
+	}
+	cl := loader.NewConfigLoader(context.Background(), sources)
+	c.loader = cl
+	c.cfg = &cfg
+	return nil
+}
+
+// LoadConfigByPath creates a config based on a path
+func (c *DynamicLoader) LoadConfigByPath(path string) error {
+	var buf []byte
+	var err error
+	if strings.HasPrefix(path, "file://") {
+		stripPath := strings.ReplaceAll(path, "file://", "")
+		buf, err = ioutil.ReadFile(stripPath)
+		if err != nil {
+			return err
+		}
+	} else if strings.HasPrefix(path, "s3://") {
+		blobURL, err := url.Parse(path)
+		if err != nil {
+			return err
+		}
+		buf, err = data.ReadBlob(*blobURL)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
nit: If the else case isn't relevant, can you add one anyway with a comment explaining why it's a no-op?


(it might be easier to read as a switch statement, but if/else if/else is fine too)

Reply:
Switched to switch with error if not s3 or file